app/src/main/java/org/thoughtcrime/securesms/profiles/AvatarHelper.java

```diff
@@ -0,0 +1,202 @@
+import org.whispersystems.signalservice.api.util.StreamDetails;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.Iterator;
+
+public class AvatarHelper {
+
+  private static final String TAG = Log.tag(AvatarHelper.class);
+
+  public static int  AVATAR_DIMENSIONS                 = 1024;
+  public static long AVATAR_DOWNLOAD_FAILSAFE_MAX_SIZE = ByteUnit.MEGABYTES.toBytes(10);
+
+  private static final String AVATAR_DIRECTORY = "avatars";
+
+  /**
+   * Retrieves an iterable set of avatars. Only intended to be used during backup.
+   */
+  public static Iterable<Avatar> getAvatars(@NonNull Context context) {
+    File   avatarDirectory = context.getDir(AVATAR_DIRECTORY, Context.MODE_PRIVATE);
+    File[] results         = avatarDirectory.listFiles();
+
+    if (results == null) {
+      return Collections.emptyList();
+    }
+
+    return () -> {
+      return new Iterator<Avatar>() {
+        int i = 0;
+        @Override
+        public boolean hasNext() {
+          return i < results.length;
+        }
+
+        @Override
+        public Avatar next() {
+          File file = results[i];
+          try {
+            return new Avatar(getAvatar(context, RecipientId.from(file.getName())),
+                              file.getName(),
+                              ModernEncryptingPartOutputStream.getPlaintextLength(file.length()));
+          } catch (IOException e) {
+            return null;
+          } finally {
+            i++;
+          }
+        }
+      };
+    };
+  }
+
+  /**
+   * Deletes and avatar.
+   */
+  public static void delete(@NonNull Context context, @NonNull RecipientId recipientId) {
+    getAvatarFile(context, recipientId).delete();
+  }
+
+  /**
+   * Whether or not an avatar is present for the given recipient.
+   */
+  public static boolean hasAvatar(@NonNull Context context, @NonNull RecipientId recipientId) {
+    File avatarFile = getAvatarFile(context, recipientId);
+    return avatarFile.exists() && avatarFile.length() > 0;
+  }
+
+  /**
+   * Retrieves a stream for an avatar. If there is no avatar, the stream will likely throw an
+   * IOException. It is recommended to call {@link #hasAvatar(Context, RecipientId)} first.
+   */
+  public static @NonNull InputStream getAvatar(@NonNull Context context, @NonNull RecipientId recipientId) throws IOException {
+    AttachmentSecret attachmentSecret = AttachmentSecretProvider.getInstance(context).getOrCreateAttachmentSecret();
+    File             avatarFile       = getAvatarFile(context, recipientId);
+
+    return ModernDecryptingPartInputStream.createFor(attachmentSecret, avatarFile, 0);
+  }
+
+  public static byte[] getAvatarBytes(@NonNull Context context, @NonNull RecipientId recipientId) throws IOException {
+    return hasAvatar(context, recipientId) ? StreamUtil.readFully(getAvatar(context, recipientId))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *getAvatar* method in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when StreamUtil.readFully throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).